internal/overlord/servstate/manager.go

```diff
@@ -53,6 +53,76 @@ func (m *ServiceManager) reloadSetup() error {
 	return nil
 }
 
+// FlattenedSetup returns the flattened setup as a single layer in YAML format.
+func (m *ServiceManager) FlattenedSetup() ([]byte, error) {
+	releaseSetup, err := m.acquireSetup()
+	if err != nil {
+		return nil, err
+	}
+	defer releaseSetup()
+
+	return m.flattened.AsYAML()
+}
+
+// MergeLayer merges the given layer YAML into the dynamic layers, returning
+// the new layer's "order" (won't increase if adding another dynamic layer).
+func (m *ServiceManager) MergeLayer(layerYAML []byte) (int, error) {
+	layer, err := setup.ParseLayer(0, "", layerYAML)
+	if err != nil {
+		return 0, err
+	}
+
+	releaseSetup, err := m.acquireSetup()
+	if err != nil {
+		return 0, err
+	}
+	defer releaseSetup()
+
+	var last *setup.Layer
+	layers := m.setup.Layers
+	if len(layers) > 0 {
+		last = layers[len(layers)-1]
+	}
+
+	var newOrder int
+	var newSetup *setup.Setup
+	var newFlattened *setup.Layer
+	if last != nil && last.IsDynamic() {
+		// Last layer is dynamic, merge new layer into existing dynamic layer
+		dynamicSetup := &setup.Setup{Layers: []*setup.Layer{last, layer}}
```

Comment:
ptr to Setup instance not necessary, also I would prefer to see Flatten become a package level function so you can just do setup.FlattenLayers(last, layer) or setup.FlattenLayers(layers...)

Reply:
You don't need a pointer here, but it's stored as a `*Setup` on the `ServiceManager` instance, so it's a bit nicer to do so (and it's also consistent with `*Layer` and the fact that the `Setup` methods are on `*Setup`). So I'm going to leave this as is.

I agree with you about `setup.FlattenLayers`, thanks -- done.